custom-api/templates/store.html

```diff
@@ -0,0 +1,279 @@
+          if (requestId !== null) {
+            setTimeout(
+              function() { 
+                pollSalesStatus(requestId, downloadSales) 
+              }, 
+              100 // 100ms polling interval.
+            )
+          }
+        });
+      });
+
+      async function downloadSales(requestId, chunkCount) {
+        var concurrencyLimit = 6; // Browser limits.
+        var mergedContentDictionary = {};
+        
+        // Create an array of indices and map them to promises for the result downloads.
+        var chunks = []
+        for (chunkIndex = 0; chunkIndex < chunkCount; chunkIndex++) {
+          chunks.push(chunkIndex)
+        }
+
+        const promises = chunks.map(function (chunkIndex) {
+          return function () {
+            return downloadFromExternalLink(requestId, chunkIndex);
+          };
+        });
+
+        // Concurrently execute the promises to download the chunks of data.
+        var requests = limitConcurrency(promises, concurrencyLimit);
+        requests
+          .then(function (responses) {
+            // Collect the unordered responses into a dictionary by chunk index.
+            for (var i = 0; i < responses.length; i++) {
+              const { chunkIndex, data } = responses[i];
+              mergedContentDictionary[chunkIndex] = data;
+            }
+
+            // Assemble the downloaded chunks in order.
+            mergedContent = "";
+            for (chunkIndex = 0; chunkIndex < chunkCount; chunkIndex++) {
+              mergedContent += mergedContentDictionary[chunkIndex];
+            }
+
+            // Create a Blob from the merged CSV data.
+            var blob = new Blob([mergedContent], { type: "text/csv" });
+
+            // Create a download link.
+            var link = document.createElement("a");
+            link.href = URL.createObjectURL(blob);
+            link.download = "sales.csv";
+
+            // Trigger a click event on the download link
+            link.click();
+            hideDownloadLoadingIcon();
+
+            // Clean up by revoking the object URL
+            URL.revokeObjectURL(link.href);
+          })
+          .catch(function (error) {
+            // Handle errors
+            console.error(error);
+          });
+      }
+
+      // Function to execute a list of promises with a concurrency limit. 
+      function limitConcurrency(promises, concurrencyLimit) {
+        const results = [];
+        let index = 0;
+
+        function next() {
+          if (index >= promises.length) {
+            return Promise.resolve();
+          }
+
+          const promise = promises[index];
+          index++;
+
+          return promise().then((result) => {
+            results.push(result);
+            return next();
+          });
+        }
+
+        const limitedPromises = Array(concurrencyLimit)
+          .fill(null)
+          .map(() => next());
+
+        return Promise.all(limitedPromises).then(() => results);
```

Comment:
does this mean we will not show all result chunks?

Reply:
This does end up getting the full set of data. I get 500MB CSV files and expect that large. There is some magic here with the promises where one completed triggers another one from the queue.